### ISSUE TITLE
Dataclass inheritance breaks default_factory field

### ISSUE THREAD

**Issue by tooncools** [pydantic/pydantic#4906]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

The use of default_factory in subclassed dataclasses invokes the TypeError "non-default argument '...' follows default argument".
Even when the subclass inherits from an empty dataclass and only the subclass contains the datafields the same TypeError occurs.
On the contrary, when creating pydantic dataclasses from stdlib dataclasses, default_factory works as expected. 

(Tested with pydantic 1.10.4 on python 3.10 and the latest 1.10.X-fixes branch that should contain the latest dataclass fixes - commit: e8934a53eba7573488f6929ca6a5c2a6d411cfdd)

### Example Code

```Python
import pydantic
from typing import List, Dict, Optional, Union, Sequence, Mapping
from dataclasses import dataclass, field

# WORKS
@dataclass
class Foo:
    a: str
    b: str = "foo"
    c: List[str] = field(default_factory=list)

pydantic.dataclasses.dataclass(Foo)


# FAILS
@dataclass
class Foo:
    a: str
    b: str = "foo"

pydantic.dataclasses.dataclass(Foo)


@dataclass
class Bar(Foo):
    c: List[str] = field(default_factory=list)

pydantic.dataclasses.dataclass(Bar)

# FAILS
@dataclass
class Foo:
    pass

pydantic.dataclasses.dataclass(Foo)


@dataclass
class Bar(Foo):
    a: str
    b: str = "foo"
    c: List[str] = field(default_factory=list)

pydantic.dataclasses.dataclass(Bar)

# WORKS
@dataclass
class Foo:
    pass

pydantic.dataclasses.dataclass(Foo)


@dataclass
class Bar(Foo):
    a: str
    b: str = "foo"
    c: str = "other"

pydantic.dataclasses.dataclass(Bar)
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.4
pydantic compiled: False
python version: 3.10.8 (main, Nov 24 2022, 14:13:03) [GCC 11.2.0]
platform: Linux-5.10.60.1-microsoft-standard-WSL2-x86_64-with-glibc2.31
optional deps. installed: ['typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [X] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by samuelcolvin** [comment]
This is not specific to pydantic, you get the same with

```py
from typing import List
from dataclasses import dataclass, field

# FAILS
@dataclass
class Foo:
    a: str
    b: str = "foo"


dataclass(Foo)


@dataclass
class Bar(Foo):
    c: List[str] = field(default_factory=list)


dataclass(Bar)
```

Solution is to use `pydantic.dataclasses.dataclass` instead of `dataclass` as the decorator.

**Comment by tooncools** [comment]
Thanks @samuelcolvin for your comment. I do feel that the use-case is a bit different: I want to "enhance" a stdlib dataclass, and create a pydantic dataclass from an inherited stdlib dataclass where we can't change the decorators. 
As found in the pydantic documentation (https://docs.pydantic.dev/usage/dataclasses/#convert-stdlib-dataclasses-into-pydantic-dataclasses), the way to do is to use the pydantic dataclass function. So rather wrapping a stdlib dataclass in a pydantic dataclass, instead of wrapping a stdlib dataclass again in an stdlib dataclass, as given by your counter-example. 

In short: the Bar dataclass is a valid dataclass (that exists perfectly fine without the non-default argument TypeError), and I want to transform it into a valid pydantic dataclass.

Is there any other way to achieve this?

Thanks for the feedback. Much appreciated.

**Comment by samuelcolvin** [comment]
Not right now, hopefully we can allow this functionality in V2 via #4669.

**Comment by tooncools** [comment]
Great, thanks for the quick response @samuelcolvin!
